### PR TITLE
Update CommandLine tests that invoke msbuild.exe to invoke it through corerun.exe

### LINKED ISSUE #261: Enable CommandLine tests that invoke msbuild.exe
Several tests invoke `msbuild.exe` to ensure that command-line handling is properly hooked up.  These tests need to be modified so that they properly invoke msbuild inside corerun if that's needed for the current architecture.  That will probably be extra exciting on Linux and OS X.

### PR DESCRIPTION
**Pull request by cdmihai**
Fixes #261 

I added a wrapper method runMSBuildWithinCoreRun that is now called by all the "MSBuild.exe" calling tests.
After talking to Daniel, it seems that in the future msbuild might be called from within other runtimes as well. The above mentioned wrapper should make it easier to generalize those tests to run in any runtime.

@AndyGerlicher @rainersigwald @dsplaisted

### CODE REVIEW

**Comment by cdmihai** [comment]
There are a couple of tests that invoke msbuild on notepad.exe just to assert for the proper errors. Notepad.exe does not exist on other OSes. Should I update those tests within this PR?

File: src/XMakeCommandLine/UnitTests/XMake_Tests.cs [hunk 1]
```diff
@@ -699,7 +698,19 @@ private void RobustDelete(string path)
             }
         }
 
-#if FEATURE_RUN_EXE_IN_TESTS
+        /// <summary>
+        /// Run msbuild.exe within corerun.exe
+        /// It concatenates the given msbuild path to the given parameters and sends it to corerun.
+        /// </summary>
+        private string runMSBuildWithinCoreRun(string pathToMsBuildExe, string msbuildParameters, bool expectSuccess = true)
```
Comment (rainersigwald) [inline src/XMakeCommandLine/UnitTests/XMake_Tests.cs]: .NET coding style is for methods to start with a capital letter, so this should be RunMSBuildWithinCoreRun.

File: src/XMakeCommandLine/UnitTests/XMake_Tests.cs [hunk 2]
```diff
@@ -847,33 +857,24 @@ public void MSBuildEngineLogger()
             }
         }
 
-        private string _pathToMSBuildExe = Path.Combine(Directory.GetCurrentDirectory(), "MSBuild.exe");
+        //private string _pathToMSBuildExe = Path.Combine(Directory.GetCurrentDirectory(), "MSBuild.exe");
```
Comment (rainersigwald) [inline src/XMakeCommandLine/UnitTests/XMake_Tests.cs]: Remove old comment

File: src/XMakeCommandLine/UnitTests/XMake_Tests.cs [hunk 3]
```diff
@@ -847,33 +857,24 @@ public void MSBuildEngineLogger()
...
         /// </summary>
         [Fact]
         public void GetCommandLine()
         {
-            string output = RunProcessAndGetOutput(_pathToMSBuildExe, "\"" + _pathToArbitraryBogusFile + "\"" + " /v:diag", expectSuccess: false);
-
-            Assert.True(output.Contains(_pathToMSBuildExe + " /v:diag " + _pathToArbitraryBogusFile));
-        }
+            var msbuildParameters = "\"" + _pathToArbitraryBogusFile + "\"" + " /v:diag";
+            Assert.True(File.Exists(_pathToArbitraryBogusFile));
```
Comment (rainersigwald) [inline src/XMakeCommandLine/UnitTests/XMake_Tests.cs]: Is this just a new sanity check?

File: src/XMakeCommandLine/UnitTests/XMake_Tests.cs [hunk 4]
```diff
@@ -891,14 +892,15 @@ public void GetCommandLineQuotedExeOnPath()
...
+                var msbuildParameters = "\"" + _pathToArbitraryBogusFile + "\"" + " /v:diag";
+                output = runMSBuildWithinCoreRun("msbuild.exe", msbuildParameters, false);
             }
             finally
             {
                 Directory.SetCurrentDirectory(current);
             }
 
-            Assert.True(output.IndexOf(_pathToMSBuildExe + " /v:diag " + _pathToArbitraryBogusFile, StringComparison.OrdinalIgnoreCase) >= 0);
+            Assert.True(output.Contains(_pathToMSBuildExe + " /v:diag " + _pathToArbitraryBogusFile));
```
Comment (rainersigwald) [inline src/XMakeCommandLine/UnitTests/XMake_Tests.cs]: You could actually change this to `Assert.Contains`.  If you didn't strip the `OrdinalIgnoreCase` on purpose, there's a bool flag for that.

File: src/XMakeCommandLine/UnitTests/XMake_Tests.cs [hunk 5]
```diff
@@ -1037,9 +1042,7 @@ public void ResponseFileInProjectDirectoryCommandLineWins()
         /// Any msbuild.rsp in the directory of the specified project/solution should be read, 
         /// but lower precedence than the actual command line and higher than the msbuild.rsp next to msbuild.exe
         /// </summary>
-        [Fact(Skip = "Ignored in MSTest")]
```
Comment (rainersigwald) [inline src/XMakeCommandLine/UnitTests/XMake_Tests.cs]: So this passes now?  Awesome!

File: src/XMakeCommandLine/UnitTests/XMake_Tests.cs [hunk 6]
```diff
@@ -633,7 +632,8 @@ public void ConfigurationInvalid()
                     "</Project>";
                 File.WriteAllText(pathToProjectFile, projectString);
 
-                output = RunProcessAndGetOutput(newPathToMSBuildExe, "\"" + pathToProjectFile + "\"", expectSuccess: false);
+                var msbuildParameters = "\"" + pathToProjectFile + "\"";
+                output = runMSBuildWithinCoreRun(newPathToMSBuildExe, msbuildParameters, expectSuccess: false);
```
Comment (rainersigwald) [inline src/XMakeCommandLine/UnitTests/XMake_Tests.cs]: Is this unconditionally running with corerun?  It should be conditional on whether we're built for coreclr--this branch should still work on full-framework (I know it doesn't currently).

**Comment by rainersigwald** [comment]
I don't think you have to fix the notepad.exe tests right now, but if you could file an issue that calls them out clearly that'd be great.

File: src/XMakeCommandLine/UnitTests/XMake_Tests.cs (hunk 3 above)
Comment (cdmihai) [inline src/XMakeCommandLine/UnitTests/XMake_Tests.cs]: Yes it is. If that file is not there, it's better for an assert to fail in a visible way than have RunProcessAndGetOutput missbehave and churn out funky state.

File: src/XMakeCommandLine/UnitTests/XMake_Tests.cs (hunk 5 above)
Comment (cdmihai) [inline src/XMakeCommandLine/UnitTests/XMake_Tests.cs]: Yup.

File: src/XMakeCommandLine/UnitTests/XMake_Tests.cs (hunk 6 above)
Comment (cdmihai) [inline src/XMakeCommandLine/UnitTests/XMake_Tests.cs]: True. I'll make a delegate out of that method that depending on an ifdef, gets instantiated with one function or the other

Comment (AndyGerlicher) [inline src/XMakeCommandLine/UnitTests/XMake_Tests.cs]: Agreed. How about rename RunMSBuildWithinCoreRun to just ExecMSBuild and have it do the right thing there based on if we need corerun or not.

Comment (cdmihai) [inline src/XMakeCommandLine/UnitTests/XMake_Tests.cs]: @AndyGerlicher I like your idea better than my delegate one. I was just doing the delegate one and all the new names made it confusing.

File: dir.props [hunk 7]
```diff
@@ -209,10 +209,10 @@
     <DefineConstants>$(DefineConstants);FEATURE_REGISTRY_SDKS</DefineConstants>
     <DefineConstants>$(DefineConstants);FEATURE_REGISTRYHIVE_DYNDATA</DefineConstants>
     <DefineConstants>$(DefineConstants);FEATURE_RESOURCE_EXPOSURE</DefineConstants>
-    <DefineConstants>$(DefineConstants);FEATURE_RUN_EXE_IN_TESTS</DefineConstants>
     <DefineConstants>$(DefineConstants);FEATURE_SECURITY_PERMISSIONS</DefineConstants>
     <DefineConstants>$(DefineConstants);FEATURE_SECURITY_PRINCIPAL_WINDOWS</DefineConstants>
     <DefineConstants>$(DefineConstants);FEATURE_SPECIAL_FOLDERS</DefineConstants>
+    <DefineConstants>$(DefineConstants);FEATURE_RUN_EXE_IN_TESTS</DefineConstants>
```
Comment (rainersigwald) [inline dir.props]: Why change the order (away from alphabetical)?

File: src/XMakeCommandLine/UnitTests/XMake_Tests.cs [hunk 8]
```diff
@@ -699,7 +698,22 @@ private void RobustDelete(string path)
             }
         }
 
+        /// <summary>
+        /// Run msbuild.exe within corerun.exe
```
Comment (rainersigwald) [inline src/XMakeCommandLine/UnitTests/XMake_Tests.cs]: Update comment for new purpose

File: dir.props (hunk 7 above)
Comment (rainersigwald) [inline dir.props]: Oh, it got removed and re-added.  Can you rearrange? :innocent:

File: src/XMakeCommandLine/UnitTests/XMake_Tests.cs [hunk 9]
```diff
@@ -699,7 +698,22 @@ private void RobustDelete(string path)
...
+        /// It concatenates the given msbuild path to the given parameters and sends it to corerun.
+        /// </summary>
+        private string ExecMSBuild(string pathToMsBuildExe, string msbuildParameters, bool expectSuccess = true)
+        {
 #if FEATURE_RUN_EXE_IN_TESTS
```
Comment (rainersigwald) [inline src/XMakeCommandLine/UnitTests/XMake_Tests.cs]: Hmm, this is kinda misnamed now.  @dsplaisted is there a good way to check for "if I need to wrap with corerun.exe"?

Comment (cdmihai) [inline src/XMakeCommandLine/UnitTests/XMake_Tests.cs]: Shouldn't it be invisible / indifferent for the method's callers how msbuild is called? In which case ExecMSBuild has the contract to "somehow" end up invoking msbuild.exe with the given parameters; how it does that should not matter.

**Comment by AndyGerlicher** [comment]
LGTM once you've addressed Rainer's comments. Thanks for rebasing the commits! :)

File: src/XMakeCommandLine/UnitTests/XMake_Tests.cs (hunk 9 above)
Comment (rainersigwald) [inline src/XMakeCommandLine/UnitTests/XMake_Tests.cs]: Yes, I agree.  I was saying that the feature flag `RUN_EXE_IN_TESTS` is now awkward:  it's really controlling whether to use corerun or not.  We need to make that decision but this seems like an inelegant way.

Comment (cdmihai) [inline src/XMakeCommandLine/UnitTests/XMake_Tests.cs]: How about `INVOKE_MSBUILD`?

Now that I'm thinking about it, I wonder if xmake_tests is the only place that invokes the msbuild.exe. I'll go check.

**Comment by rainersigwald** [comment]
LGTM

**Pull request merged**